internal/controller/sparkapplication/controller.go

```diff
@@ -328,15 +322,22 @@ func (r *Reconciler) reconcileFailedSubmissionSparkApplication(ctx context.Conte
 			app := old.DeepCopy()
 
 			if util.ShouldRetry(app) {
-				if isNextRetryDue(app) {
+				timeUntilNextRetryDue, err := util.TimeUntilNextRetryDue(app)
+				if err != nil {
+					return err
+				}
+				if timeUntilNextRetryDue <= 0 {
 					if r.validateSparkResourceDeletion(ctx, app) {
 						_ = r.submitSparkApplication(app)
 					} else {
 						if err := r.deleteSparkResources(ctx, app); err != nil {
 							logger.Error(err, "failed to delete resources associated with SparkApplication", "name", app.Name, "namespace", app.Namespace)
 						}
```

Comment:
The return error statement should be placed in the curly bracket.
```suggestion
						if err := r.deleteSparkResources(ctx, app); err != nil {
							logger.Error(err, "failed to delete resources associated with SparkApplication", "name", app.Name, "namespace", app.Namespace)
							return fmt.Errorf("resources associated with SparkApplication name: %s namespace: %s, needed to be deleted", app.Name, app.Namespace)
						}
```

Reply:
The way I had it preserves the old behaviour and I think this suggestion would actaully introduce a bug. With this suggestion, if `deleteSparkResources` succeeds then reconcile will return `ctrl.Result{}, nil`, meaning it won't be requeued. But we need it to be requeued because after `r.validateSparkResourceDeletion` returned `false` its not going to try re-submitting the the application in this reconcile. 

We could make a slightly larger change so that `submitSparkApplication` is also called if `deleteSparkResources` is successful. That would allow the submission to happen in the same reconcile, which sounds like a good thing to me but I have little experience developing kubernetes controllers.

Reply:
I recommend merging this PR in its current state. I tested it on my environment, and it fixes the `restartPolicy` issues we have been seeing.

Reply:
@Tom-Newton Yes, you are right. Will merge this PR first, we can optimize the reconcile logic later. For a failed-submission  app with max retries 3, the controller will need to reconcile 9 times now (NEW -> FAILED_SUBMISSION -> FAILED_SUBMISSION -> FAILED_SUBMISSION -> FAILED_SUBMISSION -> FAILED_SUBMISSION -> FAILED_SUBMISSION -> FAILED_SUBMISSION -> FAILED).

Reply:
I created an issue so this doesn't get forgot https://github.com/kubeflow/spark-operator/issues/2285